### PR TITLE
Add dependency to secp256k1 and add structs for keys

### PR DESCRIPTION
**Pull request by tyiu**

### CODE REVIEW

File: Sources/NostrSDK/Keys.swift [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import Foundation
+import secp256k1
+
+struct Keypair {
+    public let publicKey: PublicKey
+    public let privateKey: PrivateKey
+
+    init?() {
+        guard let key = try? secp256k1.Signing.PrivateKey() else {
+            print("Could not generate new secp256k1 private key.")
```
Comment (tyiu) [inline Sources/NostrSDK/Keys.swift:17]: Maybe it makes sense to not swallow any of the errors in this file but re-throw? What do you think?

File: Package.swift [hunk 2]
```diff
@@ -5,7 +5,6 @@ import PackageDescription
 
 let package = Package(
     name: "NostrSDK",
-    platforms: [.macOS(.v12)],
```
Comment (tyiu) [inline Package.swift:8]: This came from pulling in SwiftLint initially, which got pulled out, but we forgot to remove this line.

File: Tests/NostrSDKTests/Bech32Tests.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            } catch let error as Bech32.DecodingError {
+                XCTAssert(errorsEqual(error, reason), "Decoding error mismatch, got \(error.localizedDescription), expected \(reason.localizedDescription)")
+            } catch {
+                XCTFail("Invalid error occured: \(error.localizedDescription)")
+            }
+        }
+    }
+
+    private func errorsEqual(_ lhs: Bech32.DecodingError, _ rhs: Bech32.DecodingError) -> Bool {
+        switch lhs {
```
Comment (bryanmontz) [inline Tests/NostrSDKTests/Bech32Tests.swift:127]: There's a nicer way to do this using a tuple in the `switch`:
```swift
switch (lhs, rhs) {
case (.checksumMismatch, .checksumMismatch),
         (.incorrectChecksumSize, .incorrectChecksumSize),
        ...:
        return true 
default:
        return false
}
```

File: Tests/NostrSDKTests/Bech32Tests.swift [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        case .nonUTF8String:
+            return rhs == .nonUTF8String
+        case .stringLengthExceeded:
+            return rhs == .stringLengthExceeded
+        case .nonPrintableCharacter:
+            return rhs == .nonPrintableCharacter
+        }
+    }
+
+    static var allTests = [
```
Comment (bryanmontz) [inline Tests/NostrSDKTests/Bech32Tests.swift:149]: Remove unused code

File: Sources/NostrSDK/Keys.swift [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+//  Keys.swift
+//  
+//
+//  Created by Terry Yiu on 5/23/23.
+//
+
+import Foundation
+import secp256k1
+
+struct Keypair {
```
Comment (bryanmontz) [inline Sources/NostrSDK/Keys.swift:11]: I assume we'll want to make these `struct`s and all `init`s `public`.

File: Sources/NostrSDK/Keys.swift (hunk 1 above)
Comment (bryanmontz) [reply Sources/NostrSDK/Keys.swift:17]: If this is the kind of thing that never actually fails, maybe it's unnecessary to burden the call sites.

File: Tests/NostrSDKTests/Bech32Tests.swift [hunk 6]
```diff
@@ -0,0 +1,153 @@
+//
+//  Bech32Tests.swift
+//
+//  Modified by Terry Yiu on 5/24/23.
+//  Created by Evolution Group Ltd on 12.02.2018.
+//  Copyright © 2018 Evolution Group Ltd. All rights reserved.
```
Comment (bryanmontz) [inline Tests/NostrSDKTests/Bech32Tests.swift:6]: Are you sure this is okay to copy in to our project with an MIT license? I didn't check. Just wanting to make sure we do due diligence.

File: Tests/NostrSDKTests/Bech32Tests.swift [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        for invalid in _invalidChecksum {
+            let checksum = invalid.bech32
+            let reason = invalid.error
+            do {
+                let decoded = try Bech32.decode(checksum)
+                XCTFail("Successfully decoded an invalid checksum \(checksum): \(decoded.checksum.hexString)")
+            } catch let error as Bech32.DecodingError {
+                XCTAssert(errorsEqual(error, reason), "Decoding error mismatch, got \(error.localizedDescription), expected \(reason.localizedDescription)")
+            } catch {
+                XCTFail("Invalid error occured: \(error.localizedDescription)")
```
Comment (bryanmontz) [inline Tests/NostrSDKTests/Bech32Tests.swift:121]: Spelling: "occured" -> "occurred"

File: Tests/NostrSDKTests/Bech32Tests.swift (hunk 6 above)
Comment (tyiu) [reply Tests/NostrSDKTests/Bech32Tests.swift:6]: I sourced it from here and it's also MIT licensed. https://github.com/0xDEADP00L/Bech32

I should ask the folks from Nos, though, for some of the other code.

Comment (tyiu) [reply Tests/NostrSDKTests/Bech32Tests.swift:6]: Permission to redistribute this code under MIT license granted nostr:note1q39598qkdc093sdq4enudjf0dall76s7n779k07nutgd9r2zt6vq96l8c2

**Review by joelklabo — APPROVED** [review]

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (tyiu) [reply Package.swift:8]: I assume it supports everything by default, and you use these to scope it down.

File: Sources/NostrSDK/Data+Encoding.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+//
+//
+//  Created by Terry Yiu on 5/28/23.
+//
+
+import Foundation
+
+extension Data {
+    var hexString: String {
+        let hexDigits = Array("0123456789abcdef".utf16)
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Sources/NostrSDK/Data+Encoding.swift:14]: I copied this from Nos so I don't want to touch it as it works.

File: Tests/NostrSDKTests/Bech32Tests.swift [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+import XCTest
+@testable import NostrSDK
+
+private typealias InvalidChecksum = (bech32: String, error: Bech32.DecodingError)
+private typealias ValidAddressData = (address: String, script: [UInt8])
+private typealias InvalidAddressData = (hrp: String, version: Int, programLen: Int)
+
+class Bech32Tests: XCTestCase {
+
+    private let _validChecksum: [String] = [
```
[the remark this answers is not in the document]
Comment (tyiu) [reply Tests/NostrSDKTests/Bech32Tests.swift:22]: From @bryanmontz 

>No reason to start a property name with an underscore in Swift. I assumed it was code copied from somewhere else. In which case I’d probably just leave it.

File: Tests/NostrSDKTests/Bech32Tests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (tyiu) [reply Tests/NostrSDKTests/Bech32Tests.swift]: Came from the copied file. This test file was testing a few other things that we don't need. I deleted them but forgot to clean this up.

**Pull request merged**